File: src/integrations/data-dictionary-client.ts
/**
 * MCP client for the DataDictionary server.
 *
 * The DataDictionary server (https://github.com/rajatarun/DataDictionary) is a
 * FastMCP server deployed as an AWS Lambda function. It stores API field
 * definitions (DataElements) in DynamoDB and uses AWS Bedrock to generate
 * AI-grounded meanings.
 *
 * This client uses the MCP SDK's HTTP client to call the DataDictionary tools:
 *   - propose_data_element  → Bedrock generates a DataElement from a description
 *   - commit_data_element   → Persists the approved proposal to DynamoDB
 *   - get_data_element      → Retrieve by exact field name
 *   - search_data_elements  → Free-text search
 *   - list_data_elements    → Paginated listing with optional context filter
 *   - get_elements_by_context → All fields for a given API/service
 */

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";

export interface DataElement {
  dataElement: string;
  meaning: string;
  context: string;
  dataType: "string" | "number" | "boolean" | "array" | "object";
  examples: string[];
  constraints: string;
  relatedElements: string[];
  source: { apiName: string; version: string };
  status: "active" | "deprecated";
  createdAt?: string;
  updatedAt?: string;
}

export interface ProposeResult {
  proposal_id: string;
  commit_token?: string;
  proposed_element: DataElement;
  status: "approved" | "blocked";
  message?: string;
}

export interface SearchResult {
  items: DataElement[];
  count: number;
  last_evaluated_key?: string;
}

export class DataDictionaryClient {
  private url: string;

  constructor(dataDictionaryUrl: string) {
    // Normalise — strip trailing slash
    this.url = dataDictionaryUrl.replace(/\/$/, "");
  }

  /**
   * Call a DataDictionary MCP tool with the given arguments.
   * Creates a fresh stateless connection per call (matches server's stateless_http=True).
   */
  private async callTool(toolName: string, args: Record<string, unknown>): Promise<unknown> {
    const transport = new StreamableHTTPClientTransport(new URL(`${this.url}/mcp`));
    const client = new Client({ name: "toolweave", version: "1.0.0" });

    await client.connect(transport);
    try {
      const result = await client.callTool({ name: toolName, arguments: args });
      // Extract text content from MCP response
      const content = result.content as Array<{ type: string; text?: string }>;
      const text = content.find((c) => c.type === "text")?.text ?? "";
      try {
        return JSON.parse(text);
      } catch {
        return text;
      }
    } finally {
      await client.close();
    }
  }

  /**
   * Propose a new data element (uses Bedrock to generate structured definition).
   * Returns a proposal ID and commit token if Observatory approves.
   */
  async proposeDataElement(description: string): Promise<ProposeResult> {
    return this.callTool("propose_data_element", { description }) as Promise<ProposeResult>;
  }

  /**
   * Commit an approved proposal to DynamoDB.
   */
  async commitDataElement(proposalId: string, commitToken: string): Promise<DataElement> {
    return this.callTool("commit_data_element", {
      proposal_id: proposalId,
      commit_token: commitToken,
    }) as Promise<DataElement>;
  }

  /**
   * Retrieve a data element by its exact field name.
   */
  async getDataElement(dataElement: string): Promise<DataElement | null> {
    try {
      return (await this.callTool("get_data_element", { data_element: dataElement })) as DataElement;
    } catch {
      return null;
    }
  }

  /**
   * Free-text search across field names and meanings.
   */
  async searchDataElements(query: string, limit = 20): Promise<SearchResult> {
    return this.callTool("search_data_elements", { query, limit }) as Promise<SearchResult>;
  }

  /**
   * List data elements, optionally filtered by context (API name).
   */
  async listDataElements(context?: string, limit = 50): Promise<SearchResult> {
    return this.callTool("list_data_elements", { context, limit }) as Promise<SearchResult>;
  }

  /**
   * Get all data elements for a given API/service context.
   */
  async getElementsByContext(context: string): Promise<DataElement[]> {
    const result = (await this.callTool("get_elements_by_context", {
      context,
    })) as { items: DataElement[] } | DataElement[];
    return Array.isArray(result) ? result : result.items ?? [];
  }
}

/**
 * Build a natural language description for a field to send to propose_data_element.
 * The DataDictionary's Bedrock prompt works best with rich context.
 */
export function buildFieldDescription(opts: {
  fieldName: string;
  fieldType: string;
  description?: string;
  apiName: string;
  apiVersion: string;
  endpointPath: string;
  endpointMethod: string;
  required: boolean;
  example?: unknown;
  enumValues?: unknown[];
  format?: string;
}): string {
  const parts: string[] = [
    `Field name: "${opts.fieldName}"`,
    `API: "${opts.apiName}" (version ${opts.apiVersion})`,
    `Endpoint: ${opts.endpointMethod} ${opts.endpointPath}`,
    `Data type: ${opts.fieldType}${opts.format ? ` (format: ${opts.format})` : ""}`,
    `Required: ${opts.required ? "yes" : "no"}`,
  ];

  if (opts.description) parts.push(`Description from spec: ${opts.description}`);
  if (opts.example !== undefined) parts.push(`Example value: ${JSON.stringify(opts.example)}`);
  if (opts.enumValues?.length) parts.push(`Allowed values: ${opts.enumValues.join(", ")}`);

  return parts.join(". ");
}

File: src/toolweave/data_dictionary_client.py
from __future__ import annotations

import json
import os
from typing import Any, Optional

import httpx

from .models import DataElementMeta

DATA_DICTIONARY_URL = os.environ.get(
    "DATA_DICTIONARY_URL",
    "https://ft5yitoykshqz73sejd7miatze0syqds.lambda-url.us-east-1.on.aws/",
)

# MCP endpoint — the DataDictionary Lambda uses stateless_http=True so every
# POST to /mcp is a self-contained request/response session.
_MCP_ENDPOINT = DATA_DICTIONARY_URL.rstrip("/") + "/mcp"


async def _call_tool(tool_name: str, arguments: dict[str, Any]) -> Any:
    """Send a single MCP tool call to the DataDictionary Lambda and return the result."""
    payload = {
        "jsonrpc": "2.0",
        "id": 1,
        "method": "tools/call",
        "params": {
            "name": tool_name,
            "arguments": arguments,
        },
    }
    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            _MCP_ENDPOINT,
            json=payload,
            headers={"Content-Type": "application/json"},
        )
        resp.raise_for_status()
        body = resp.json()

    # MCP response: {"result": {"content": [{"type": "text", "text": "..."}]}}
    result = body.get("result", {})
    content = result.get("content", [])
    if content and content[0].get("type") == "text":
        text = content[0]["text"]
        try:
            return json.loads(text)
        except (json.JSONDecodeError, ValueError):
            return text
    return result


async def get_data_element(data_element: str) -> Optional[dict[str, Any]]:
    result = await _call_tool("get_data_element", {"dataElement": data_element})
    if isinstance(result, dict) and "error" not in result:
        return result
    return None


async def search_data_elements(query: str) -> list[dict[str, Any]]:
    result = await _call_tool("search_data_elements", {"query": query})
    if isinstance(result, list):
        return result
    return []


async def get_elements_by_context(context: str) -> list[dict[str, Any]]:
    result = await _call_tool("get_elements_by_context", {"context": context})
    if isinstance(result, list):
        return result
    return []


async def fetch_field_metadata(
    field_names: list[str],
    context: str = "",
) -> dict[str, Optional[DataElementMeta]]:
    """Fetch DataDictionary metadata for a list of field names.

    Strategy:
    1. Try exact `get_data_element` lookup per field name
    2. Fallback: `search_data_elements` for fields not found
    3. Fallback: bulk `get_elements_by_context` filtered by name

    Returns a mapping {field_name: DataElementMeta | None}.
    """
    result: dict[str, Optional[DataElementMeta]] = {n: None for n in field_names}
    missing: list[str] = list(field_names)

    # Pass 1 — exact lookup
    for name in list(missing):
        item = await get_data_element(name)
        if item:
            try:
                result[name] = DataElementMeta(**item)
                missing.remove(name)
            except Exception:
                pass

    if not missing:
        return result

    # Pass 2 — search lookup for remaining fields
    still_missing: list[str] = []
    for name in missing:
        items = await search_data_elements(name)
        # Pick best match: exact name match first, otherwise first result
        matched = next((i for i in items if i.get("dataElement") == name), None)
        if matched is None and items:
            matched = items[0]
        if matched:
            try:
                result[name] = DataElementMeta(**matched)
            except Exception:
                still_missing.append(name)
        else:
            still_missing.append(name)

    if not still_missing or not context:
        return result

    # Pass 3 — bulk context fetch
    all_items = await get_elements_by_context(context)
    by_name = {i.get("dataElement", ""): i for i in all_items}
    for name in still_missing:
        item = by_name.get(name)
        if item:
            try:
                result[name] = DataElementMeta(**item)
            except Exception:
                pass

    return result
